File: job-2022-autumn/40.combination-sum-ii.js
/*
 * @lc app=leetcode id=40 lang=javascript
 *
 * [40] Combination Sum II
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function (candidates, target) {
    candidates.sort((a, b) => a - b);

    /** @type {number[][]} */
    const results = [];
    /** @type {number[]} */
    const currentResult = [];
    let currentSum = 0;

    /**
     * @param {number} startIndex
     */
    const backtrack = (startIndex) => {
        if (currentSum === target) {
            results.push([...currentResult]);
        } else if (currentSum + candidates[startIndex] > target) {
            return;
        } else {
            for (let i = startIndex; i < candidates.length; i++) {
                if (i === startIndex || candidates[i] !== candidates[i - 1]) {
                    currentResult.push(candidates[i]);
                    currentSum += candidates[i];
                    backtrack(i + 1);
                    currentSum -= candidates[i];
                    currentResult.pop();
                }
            }
        }
    };

    backtrack(0);

    return results;
};
// @lc code=end

File: Daily/40.combination-sum-ii.cpp
/*
 * @lc app=leetcode id=40 lang=cpp
 *
 * [40] Combination Sum II
 */
#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  std::vector<std::vector<int>> combinationSum2(std::vector<int>& candidates,
                                                const int target) {
    std::sort(candidates.begin(), candidates.end());
    backtrack(candidates, 0, target);
    return resultCombinations;
  }

 private:
  void backtrack(const std::vector<int>& candidates,
                 const int index,
                 const int target) {
    if (currentSum > target) {
      return;
    }

    if (currentSum == target) {
      resultCombinations.push_back(currentCombination);
      return;
    }

    for (int i = index; i < candidates.size(); i++) {
      if (i > 0 && i != index && candidates[i - 1] == candidates[i]) {
        continue;
      }
      currentSum += candidates[i];
      currentCombination.push_back(candidates[i]);
      backtrack(candidates, i + 1, target);
      currentCombination.pop_back();
      currentSum -= candidates[i];
    }
  }

 private:
  int currentSum = 0;
  std::vector<int> currentCombination;
  std::vector<std::vector<int>> resultCombinations;
};
// @lc code=end

int main() {
  Solution sol;
  std::vector<int> nums = {10, 1, 2, 7, 6, 1, 5};
  sol.combinationSum2(nums, 8);
}
